In [0]:
import numpy as np
import pandas as pd
from sklearn import svm
from sklearn.ensemble import AdaBoostClassifier
from sklearn.metrics import confusion_matrix

In [0]:
# train from example 03
X_train_a = pd.read_csv("example03-data.csv",header=None)
y_train_a = pd.read_csv("example03-class.csv", header=None) 

The test sets will be half of example04 for all models. <br>
Preparing them ahead.

In [0]:
# example 04
X = pd.read_csv("example04-data.csv", header=None)
y = pd.read_csv("example04-class.csv", header=None)

In [0]:
# splitting half
from sklearn.model_selection import train_test_split
X_train_b, X_test, y_train_b, y_test = train_test_split(
    X, y, test_size=0.5, random_state=42)

In [0]:
# make array

# a - eg3
X_train_a = np.asarray(X_train_a)
y_train_a = np.asarray(y_train_a)

# b - 50% of eg4
X_train_b = np.asarray(X_train_b)
y_train_b = np.asarray(y_train_b)

# shared - 50% of eg4
X_test = np.asarray(X_test)
y_test = np.asarray(y_test)

### A

#### Decision Tree

In [0]:
# boosted decision classifier
mod1 = AdaBoostClassifier(n_estimators= 50, random_state=0)

In [7]:
# train
mod1.fit(X_train_a, y_train_a)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None, learning_rate=1.0,
                   n_estimators=50, random_state=0)

In [0]:
# predict
y_pred = mod1.predict(X_test)

In [9]:
# Plot confusion matrix
cm = confusion_matrix(y_test,y_pred, labels=[1,-1])

total = len(y_test)

# from confusion matrix calculate accuracy
accuracy=(cm[0,0]+cm[1,1])/total
print ('Accuracy : ', accuracy)

sensitivity = cm[0,0]/(cm[0,0]+cm[0,1])
print('Sensitivity : ', sensitivity)

specificity = cm[1,1]/(cm[1,0]+cm[1,1])
print('Specificity : ', specificity)

Accuracy :  0.6
Sensitivity :  0.6666666666666666
Specificity :  0.5


#### Support Vector Machine

In [0]:
# support vector classifier
mod2 = AdaBoostClassifier(svm.SVC(gamma="scale"), algorithm="SAMME",
                         n_estimators=50, random_state=0, learning_rate=1.0)

In [11]:
# train
mod2.fit(X_train_a, y_train_a)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostClassifier(algorithm='SAMME',
                   base_estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                                      class_weight=None, coef0=0.0,
                                      decision_function_shape='ovr', degree=3,
                                      gamma='scale', kernel='rbf', max_iter=-1,
                                      probability=False, random_state=None,
                                      shrinking=True, tol=0.001,
                                      verbose=False),
                   learning_rate=1.0, n_estimators=50, random_state=0)

In [0]:
# predict
y_pred = mod2.predict(X_test)

In [13]:
# Plot confusion matrix
cm = confusion_matrix(y_test,y_pred, labels=[1,-1])

total = len(y_test)

# from confusion matrix calculate accuracy
accuracy=(cm[0,0]+cm[1,1])/total
print ('Accuracy : ', accuracy)

sensitivity = cm[0,0]/(cm[0,0]+cm[0,1])
print('Sensitivity : ', sensitivity)

specificity = cm[1,1]/(cm[1,0]+cm[1,1])
print('Specificity : ', specificity)

Accuracy :  0.6
Sensitivity :  0.6666666666666666
Specificity :  0.5


### B

#### Decision Tree

In [0]:
# boosted decision classifier
mod3 = AdaBoostClassifier(n_estimators= 50, random_state=0)

In [15]:
# train
mod3.fit(X_train_b, y_train_b)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostClassifier(algorithm='SAMME.R', base_estimator=None, learning_rate=1.0,
                   n_estimators=50, random_state=0)

In [0]:
# predict
y_pred = mod3.predict(X_test)

In [17]:
# Plot confusion matrix
cm = confusion_matrix(y_test,y_pred, labels=[1,-1])

total = len(y_test)

# from confusion matrix calculate accuracy
accuracy=(cm[0,0]+cm[1,1])/total
print ('Accuracy : ', accuracy)

sensitivity = cm[0,0]/(cm[0,0]+cm[0,1])
print('Sensitivity : ', sensitivity)

specificity = cm[1,1]/(cm[1,0]+cm[1,1])
print('Specificity : ', specificity)

Accuracy :  0.2
Sensitivity :  0.0
Specificity :  0.5


#### Support Vector Machine

In [0]:
# support vector classifier
mod4 = AdaBoostClassifier(svm.SVC(gamma="scale"), algorithm="SAMME",
                         n_estimators=50, random_state=0, learning_rate=1.0)

In [19]:
# train
mod4.fit(X_train_b, y_train_b)

/usr/local/lib/python3.6/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


AdaBoostClassifier(algorithm='SAMME',
                   base_estimator=SVC(C=1.0, break_ties=False, cache_size=200,
                                      class_weight=None, coef0=0.0,
                                      decision_function_shape='ovr', degree=3,
                                      gamma='scale', kernel='rbf', max_iter=-1,
                                      probability=False, random_state=None,
                                      shrinking=True, tol=0.001,
                                      verbose=False),
                   learning_rate=1.0, n_estimators=50, random_state=0)

In [0]:
# predict
y_pred = mod4.predict(X_test)

In [21]:
# Plot confusion matrix
cm = confusion_matrix(y_test,y_pred, labels=[1,-1])

total = len(y_test)

# from confusion matrix calculate accuracy
accuracy=(cm[0,0]+cm[1,1])/total
print ('Accuracy : ', accuracy)

sensitivity = cm[0,0]/(cm[0,0]+cm[0,1])
print('Sensitivity : ', sensitivity)

specificity = cm[1,1]/(cm[1,0]+cm[1,1])
print('Specificity : ', specificity)

Accuracy :  0.4
Sensitivity :  0.0
Specificity :  1.0


#### Comparison between Models

Model | Train Set | Train Size | Accuracy | Sensitivity | Specificity
-|-|-|-|-|-
 Decision Tree | example03 | 10 | 0.60 | 0.67 | 0.5
 Support Vector Classifier | example03 | 10 | 0.60 | 0.67 | 0.5
 Decision Tree | example04 | 5 | 0.20 | 0.0 | 0.5
 Support Vector Classifier | example04 | 5 | 0.40 | 0.0 | 1.0